### PR TITLE
Add an operator Helm chart

### PR DESCRIPTION
**Pull request by skitt**
The submariner-operator Helm chart deploys Submariner using the
Submariner operator: it deploys the CRDs, roles, and SAs required to
run the operator and Submariner, then pushes a Submariner CR to
instruct the operator to deploy Submariner.

This patch also adds a Makefile to validate the deployments; run

	make deploy

to test the Helm charts (this will bring up two clusters, install the
broker on one of them and the operator on both, and end with a
connectivity test).

Fixes: #17
Signed-off-by: Stephen Kitt <skitt@redhat.com>

### CODE REVIEW

**Comment by skitt** [comment]
This is still full of cruft; but it should allow deployments with the operator.

**Comment by mangelajo** [comment]
I'm all in! ;)

File: .travis.yml [hunk 1]
```diff
@@ -18,15 +18,15 @@ install:
   - sudo pip install yamllint=="${YAMLLINT_VERSION}"
 
 script:
-  - for dir in submariner submariner-k8s-broker; do helm lint $dir $HELM_FLAGS; done
+  - for dir in submariner submariner-k8s-broker submariner-operator; do helm lint $dir $HELM_FLAGS; done
```
Comment (dfarrell07) [inline .travis.yml:21]: Can we make a parallel change to https://github.com/submariner-io/submariner-charts/blob/master/.github/workflows/linting.yml#L56?

Comment (dfarrell07) [reply .travis.yml:21]: Thanks @skitt!

**Comment by dfarrell07** [comment]
I added some basic `make deploy` CI in a PR based on this. It works! :smile: https://github.com/submariner-io/submariner-charts/pull/72

**Review by dfarrell07 — APPROVED** [review]

**Review by tpantelis — APPROVED** [review]

**Comment by dfarrell07** [comment]
Ping for review @mangelajo :pray:

File: deploy_helm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        --set submariner.natEnabled="false" \
+        --set routeAgent.image.repository="localhost:5000/submariner-route-agent" \
+        --set routeAgent.image.tag="local" \
+        --set routeAgent.image.pullPolicy="IfNotPresent" \
+        --set engine.image.repository="localhost:5000/submariner" \
+        --set engine.image.tag="local" \
+        --set engine.image.pullPolicy="IfNotPresent" \
+        --set globalnet.image.repository="localhost:5000/submariner-globalnet" \
+        --set globalnet.image.tag="local" \
+        --set globalnet.image.pullPolicy="IfNotPresent" \
```
Comment (mangelajo) [inline deploy_helm:81]: do those matter anymore if we are deploying with the operator?

File: deploy_helm [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    submariner_broker_url=$(kubectl -n default get endpoints kubernetes -o jsonpath="{.subsets[0].addresses[0].ip}:{.subsets[0].ports[?(@.name=='https')].port}")
+    submariner_broker_ca=$(kubectl -n "${SUBMARINER_BROKER_NS}" get secrets -o jsonpath="{.items[?(@.metadata.annotations['kubernetes\.io/service-account\.name']=='${SUBMARINER_BROKER_NS}-client')].data['ca\.crt']}")
+    submariner_broker_token=$(kubectl -n "${SUBMARINER_BROKER_NS}" get secrets -o jsonpath="{.items[?(@.metadata.annotations['kubernetes\.io/service-account\.name']=='${SUBMARINER_BROKER_NS}-client')].data.token}"|base64 --decode)
+}
+
+function helm_install_subm() {
+    local crd_create=false
+    [[ "${cluster}" = "${broker}" ]] || crd_create=true
+
+    if kubectl wait --for=condition=Ready pods -l app=submariner-engine -n "${SUBM_NS}" --timeout=60s > /dev/null 2>&1; then
```
Comment (mangelajo) [inline deploy_helm:52]: nit: we could wait for the operator instead.

**Review by mangelajo — APPROVED** [review]
Just a couple of nits which we could handle in follow-up (if the nits are correct).

Also, at this point we can delete the "submariner" chat, keeping only submariner-operator & submariner-k8s-broker. Right?

File: deploy_helm [hunk 4]
```diff
@@ -0,0 +1,88 @@
+# shellcheck shell=bash
```
Comment (mkolesnik) [inline deploy_helm:1]: I guess this has some modifications and once it all works you'll port this to Shipyard?

File: submariner-operator/templates/crd.yaml [hunk 5]
```diff
@@ -0,0 +1,723 @@
+---
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (mkolesnik) [inline submariner-operator/templates/crd.yaml:2]:
```suggestion
apiVersion: apiextensions.k8s.io/v1
```

No?

**Review by mkolesnik — DISMISSED** [review]

**Comment by skitt** [comment]
> Also, at this point we can delete the "submariner" chat, keeping only submariner-operator & submariner-k8s-broker. Right?

Yes, I’ve got that waiting in the wings, but I want to get this merged, update the docs, and only then drop the old chart.

File: deploy_helm (hunk 4 above)
Comment (skitt) [reply deploy_helm:1]: Yes, the idea was to test them here first, and also validate that the patch works.

File: submariner-operator/templates/crd.yaml (hunk 5 above)
Comment (skitt) [reply submariner-operator/templates/crd.yaml:2]: No, the v1 schema is different, we’ll migrate all the CRDs simultaneously once we get them generated properly.

File: deploy_helm (hunk 2 above)
Comment (skitt) [reply deploy_helm]: Most of them don’t, I’ve cleaned them up.

File: deploy_helm (hunk 3 above)
Comment (skitt) [reply deploy_helm]: Indeed, done!

File: cluster_settings [hunk 6]
```diff
@@ -0,0 +1,10 @@
+. "${SCRIPTS_DIR}"/lib/source_only
+
+# We need a minimal setup to verify the deployment works
+clusters=('cluster1' 'cluster2')
+cluster_nodes['cluster1']="control-plane worker worker"
+cluster_nodes['cluster2']="control-plane worker worker"
```
[the remark this answers is not in the document]
Comment (skitt) [reply cluster_settings]: Yup, I’ve dropped the extra workers.

**Comment by skitt** [comment]
@mkolesnik could you check whether the current version addresses your concerns? We can’t merge this as long as your comments aren’t addressed.

**Comment by mangelajo** [comment]
@skitt I see the comment from mike but you both seem to be agreeing. I think it's fine to merge as long as we move it to shipyard afterwards

**Review by mangelajo — APPROVED** [review]

**Comment by mangelajo** [comment]
I see the changes requested are done or makes sense to handle as follow ups (as they are planned)

**Pull request merged**